### PR TITLE
hwloc : support for v2.0.0rc2

### PR DESCRIPTION
**Pull request by KavithaTipturMadhu**
changes for hwloc's newer version, requires that netloc build references be removed. working on exposing netloc as a library. need the code changes to be reviewed in the meantime.

### CODE REVIEW

File: configure.ac [hunk 1]
```diff
@@ -1467,7 +1467,7 @@ if test "$with_hwloc_prefix" = "embedded" ; then
          CFLAGS="$HWLOC_EMBEDDED_VISIBILITY_CFLAGS $CFLAGS"
       fi
       hwlocsrcdir="src/hwloc"
-      hwloclib="src/hwloc/src/libhwloc_embedded.la"
+      hwloclib="src/hwloc/hwloc/libhwloc_embedded.la"
```
Comment (raffenet) [inline configure.ac]: Using `$HWLOC_EMBEDDED_LDADD` should make it so we don't have to manually discover the path.

**Comment by pavanbalaji** [comment]
This patch is mixing the upgrade to `hwloc-2.0` with other unrelated changes in the code.  Those need to be separated into different patches.

**Comment by KavithaTipturMadhu** [comment]
@pavanbalaji could you please point out which changes need to be moved to a separate patch?

File: configure.ac (hunk 1 above)
Comment (KavithaTipturMadhu) [reply configure.ac]: will fix this

**Comment by KavithaTipturMadhu** [comment]
@wesbland yes, I'm working on exposing netloc too, will create a patch for the submodule.

File: src/mpi/comm/comm_split_type.c [hunk 2]
```diff
@@ -106,79 +106,47 @@ int MPIR_Comm_split_type_node(MPIR_Comm * user_comm_ptr, int split_type, int key
...
-    union {
-        struct {
-            hwloc_obj_cache_type_t cache_type;
-            int depth;
-        } cache;
-    } u;
-};
-
-struct shmem_processor_info_table {
+struct shmem_info_table {
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: This change seems orthogonal to the `hwloc-2.0` upgrade.

File: src/mpi/comm/comm_split_type.c [hunk 3]
```diff
@@ -106,79 +106,47 @@ int MPIR_Comm_split_type_node(MPIR_Comm * user_comm_ptr, int split_type, int key
...
     const char *val;
     hwloc_obj_type_t obj_type;
 };
 
 /* hwloc processor object table */
-static struct shmem_processor_info_table shmem_processor_info[] = {
+static struct shmem_info_table shmem_info[] = {
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: Unrelated to the `hwloc-2.0` upgrade.

File: src/mpi/comm/comm_split_type.c [hunk 4]
```diff
@@ -195,8 +163,7 @@ static int node_split_processor(MPIR_Comm * comm_ptr, int key, struct info_obj i
         hwloc_get_obj_covering_cpuset(MPIR_Process.topology, MPIR_Process.bindset);
     if (obj_containing_cpuset->type == query_obj_type) {
         color = obj_containing_cpuset->logical_index;
-    }
-    else {
+    } else {
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: White space change.  Unrelated to the `hwloc-2.0` upgrade.

**Comment by pavanbalaji** [comment]
I've listed a few examples above.  But you should look at each of your changed lines and figure out if it is related or not.

File: src/mpi/comm/comm_split_type.c (hunk 2 above)
Comment (KavithaTipturMadhu) [reply src/mpi/comm/comm_split_type.c]: sorry, will fix this.

File: src/mpi/comm/comm_split_type.c [hunk 5]
```diff
@@ -130,55 +120,33 @@ static struct shmem_processor_info_table shmem_processor_info[] = {
...
+    {"cache:l3", HWLOC_OBJ_L1CACHE},
+    {"cache:l4d", HWLOC_OBJ_L1CACHE},
+    {"cache:l4u", HWLOC_OBJ_L1CACHE},
+    {"cache:l4", HWLOC_OBJ_L1CACHE},
+    {"cache:l5d", HWLOC_OBJ_L1CACHE},
+    {"cache:l5u", HWLOC_OBJ_L1CACHE},
+    {"cache:l5", HWLOC_OBJ_L1CACHE},
+    {"cache:l1i", HWLOC_OBJ_L1ICACHE},
+    {"cache:l2i", HWLOC_OBJ_L2ICACHE},
+    {"cache:l3i", HWLOC_OBJ_L3ICACHE},
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: Can you order them and make this consistent?

File: src/mpi/comm/comm_split_type.c [hunk 6]
```diff
@@ -253,67 +222,37 @@ int MPIR_Comm_split_type_node_topo(MPIR_Comm * user_comm_ptr, int split_type, in
...
-                    if (*d == 'u')
-                        info_obj.u.cache.cache_type = HWLOC_OBJ_CACHE_UNIFIED;
-                    else if (*d == 'i')
-                        info_obj.u.cache.cache_type = HWLOC_OBJ_CACHE_INSTRUCTION;
-                    else if (*d == 'd' || *d == '\0')
-                        info_obj.u.cache.cache_type = HWLOC_OBJ_CACHE_DATA;
-                    else
-                        goto use_node_comm;
-                }
-            }
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: It's OK to remove this for now, but once we add the I/O devices, we'll need to revive some part of this code to deal with them.

File: src/mpi/comm/comm_split_type.c [hunk 7]
```diff
@@ -330,15 +269,10 @@ int MPIR_Comm_split_type_node_topo(MPIR_Comm * user_comm_ptr, int split_type, in
...
+    if (obj_type == HWLOC_OBJ_TYPE_MAX)
         goto use_node_comm;
-
-    if (info_obj.obj_type == HWLOC_OBJ_CACHE) {
-        mpi_errno = node_split_cache(comm_ptr, key, info_obj, newcomm_ptr);
-    }
-    else {
-        mpi_errno = node_split_processor(comm_ptr, key, info_obj, newcomm_ptr);
-    }
+    else
```
Comment (pavanbalaji) [inline src/mpi/comm/comm_split_type.c]: Don't need this "else" here.

File: src/pm/hydra/tools/topo/hwloc/Makefile.mk [hunk 8]
```diff
@@ -15,6 +15,6 @@ AM_CPPFLAGS += -I$(top_srcdir)/tools/topo/hwloc/hwloc/include \
 # Append hwloc to the external subdirs, so it gets built first
 external_subdirs += tools/topo/hwloc/hwloc
 external_dist_subdirs += tools/topo/hwloc/hwloc
-external_ldflags += -L$(top_builddir)/tools/topo/hwloc/hwloc/src
+external_ldflags += -L$(top_builddir)/tools/topo/hwloc/hwloc/hwloc
```
Comment (pavanbalaji) [inline src/pm/hydra/tools/topo/hwloc/Makefile.mk]: I believe this goes away with Ken's PR.

File: src/pm/hydra/tools/topo/hwloc/topo_hwloc.c [hunk 9]
```diff
@@ -424,9 +424,10 @@ HYD_status HYDT_topo_hwloc_init(const char *binding, const char *mapping, const
     else if (!strncmp(membind, "interleave:", strlen("interleave:"))) {
         HYDT_topo_hwloc_info.membind = HWLOC_MEMBIND_INTERLEAVE;
     }
-    else if (!strncmp(membind, "replicate:", strlen("replicate:"))) {
+    /*else if (!strncmp(membind, "replicate:", strlen("replicate:"))) {
         HYDT_topo_hwloc_info.membind = HWLOC_MEMBIND_REPLICATE;
-    }
+  	}
+  	*/
```
Comment (pavanbalaji) [inline src/pm/hydra/tools/topo/hwloc/topo_hwloc.c]: Please just delete this instead of commenting it out.  That's what version control systems are for.

File: configure.ac [hunk 10]
```diff
@@ -1467,7 +1467,7 @@ if test "$with_hwloc_prefix" = "embedded" ; then
          CFLAGS="$HWLOC_EMBEDDED_VISIBILITY_CFLAGS $CFLAGS"
       fi
       hwlocsrcdir="src/hwloc"
-      hwloclib="src/hwloc/src/libhwloc_embedded.la"
+      hwloclib=$HWLOC_EMBEDDED_LDADD
```
Comment (pavanbalaji) [inline configure.ac]: This should be rebased on top of Ken's PR.

File: src/pm/hydra/tools/topo/hwloc/topo_hwloc.c [hunk 11]
```diff
@@ -424,9 +424,6 @@ HYD_status HYDT_topo_hwloc_init(const char *binding, const char *mapping, const
     else if (!strncmp(membind, "interleave:", strlen("interleave:"))) {
         HYDT_topo_hwloc_info.membind = HWLOC_MEMBIND_INTERLEAVE;
     }
-    else if (!strncmp(membind, "replicate:", strlen("replicate:"))) {
-        HYDT_topo_hwloc_info.membind = HWLOC_MEMBIND_REPLICATE;
-    }
```
Comment (pavanbalaji) [inline src/pm/hydra/tools/topo/hwloc/topo_hwloc.c]: Removing an option in Hydra should be a separate patch.  Was this removed from the help message too?

**Comment by pavanbalaji** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/most

**Comment by pavanbalaji** [comment]
This fails on Solaris right now.  The issue has been reported to `hwloc` upstream and should be fixed in `hwloc-2.0.0rc2`.  This PR needs to wait till that's available and we upgrade to it.  I've marked it as "do-not-merge" for now.

**Comment by KavithaTipturMadhu** [comment]
Problem is with strncasecmp function used in solaris that happens to be an extern function in strings.h header. When compiling with -Werror=implicit-function-definitio
n flag, this throws an error in place of a warning. This can be fixed by using a fallback function in place of strncasecmp for solaris in hwloc.

**Comment by pavanbalaji** [comment]
`strncasecmp` is already present on Solaris.  There's no need to do that or for any overcomplicated solution.  Please wait for `hwloc-2.0.0rc2` to be available.

**Comment by KavithaTipturMadhu** [comment]
strncasecmp is declared as an extern function in strings.h. Including the header leads to a build error when compiling with -Werror=implicit-function-declaration since the function declaration is extern.

**Comment by pavanbalaji** [comment]
`strncasecmp` works fine on Solaris.  It's just some simple header ordering issue in `hwloc`.  Can you please just wait for `hwloc-2.0.0rc2`?

**Comment by pavanbalaji** [comment]
Force pushed upgrade to `hwloc-2.0.0rc2`.  Restarting the tests now.

test:jenkins/ch4/most
test:jenkins/ch3/most

**Comment by KavithaTipturMadhu** [comment]
can this pr be merged?

**Comment by pavanbalaji** [comment]
Doesn't Hydra need to be updated to match what's needed in `hwloc-2.0`?
